lib/kino/js/live/context.ex

```diff
@@ -0,0 +1,58 @@
+defmodule Kino.JS.Live.Context do
+  @moduledoc """
+  State available in `Kino.JS.Live` server callbacks.
+  """
+
+  defstruct [:assigns, :events]
+
+  @type t :: %__MODULE__{assigns: map(), events: list()}
+
+  @doc false
+  def new() do
+    %__MODULE__{assigns: %{}, events: []}
+  end
+
+  @doc """
+  Stores key-value pairs in the state.
+
+  ## Examples
+
+      assign(ctx, count: 1, timestamp: DateTime.utc_now())
+  """
+  @spec assign(t(), Enumerable.t()) :: t()
+  def assign(%__MODULE__{} = ctx, assigns) do
+    assigns =
+      Enum.reduce(assigns, ctx.assigns, fn {key, val}, assigns ->
+        Map.put(assigns, key, val)
+      end)
+
+    %{ctx | assigns: assigns}
+  end
+
+  @doc """
+  Updates an existing key with the given function in the state.
+
+  ## Examples
+
+      assign(ctx, count: 1, timestamp: DateTime.utc_now())
+  """
+  @spec update(t(), term(), (term() -> term())) :: t()
+  def update(%__MODULE__{} = ctx, key, fun) when is_function(fun, 1) do
+    val = Map.fetch!(ctx.assigns, key)
+    assign(ctx, [{key, fun.(val)}])
+  end
+
+  @doc """
+  Sends an event to the client.
+
+  The event is dispatched to the registered JavaScript callback
+  on all connected clients.
+
+  ## Examples
+
+      broadcast_event(ctx, "new_point", %{x: 10, y: 10})
+  """
+  def broadcast_event(%__MODULE__{} = ctx, event, payload \\ nil) when is_binary(event) do
+    update_in(ctx, [Access.key!(:events)], &[{event, payload} | &1])
```

Comment:
Is there any reason to have this? Instead we could push the event directly the moment broadcast is called. That's how Phoenix Channels work. LiveView do not work like that because they always need to render.

Reply:
Given that `Context` is exposed to the user, it seems cleaner to store events there than private information like `client_pids`. Either approach works and I kinda like this more, any reason to not do that? (also that's something we can always change without affecting the API)

Reply:
The current API makes the cases where you need to do multiple sends harder. Imagine you want to send an initial payload, compute something, and send another payload. The only way to do so is by returning from the callback, send myself a message, and handling it in another callback.

Reply:
I see, thanks!